Add SwitchBee Integration

jafar-atili opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 
 https://github.com/home-assistant/home-assistant.io/pull/24019

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [x] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

jafar-atili committed:
Add SwitchBee Integration

jafar-atili committed:
Add SwitchBee Integration

homeassistant commented:
Hi @jafar-atili, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#70201).

Once you do that we will be able to review and accept this pull request.

Thanks!

jafar-atili commented:
Just Signed, I see the CI failed on importing a module from the  helper library, what did I miss?

jafar-atili committed:
fixes

jafar-atili committed:
fixes

thecode commented:
Hi @jafar-atili,
Thanks for your contribution.
The checkboxes on the PR description are there to make sure you followed the steps:
![image](https://user-images.githubusercontent.com/1858925/163791121-f026a334-b247-4952-9757-4e7064dc8976.png)

I highly suggest to setup development environment following the instructions here: https://developers.home-assistant.io/docs/development_environment
With the development environment most of the CI steps are running locally before commit.

Thanks 👍

jafar-atili commented:
> I highly suggest to setup development environment following the instructions here: https://developers.home-assistant.io/docs/development_environment

Thanks for your reply
My development environment is set on a devcontainer using docker and VSC following the instruction you mentioned.
However I struggled at the beginning because of the docker came with early version of python 3 and many dependency were not working, I had to manually build python3.9 and install all the deps - hope there's a easy solution for this.

second question, say I fixed some issue on the code, do I have to push another commit? can't I just ammend a patchset to the PR (like Gerrit)?? --- sorry but I am new to github :)

thecode commented:
> Thanks for your reply My development environment is set on a devcontainer using docker and VSC following the instruction you mentioned. However I struggled at the beginning because of the docker came with early version of python 3 and many dependency were not working, I had to manually build python3.9 and install all the deps - hope there's a easy solution for this.

The dev container comes with it's own python, you don't need to install python on the host system. It looks like something did not work for you on the initial installation. 

Try running `pre-commit`, if it doesn't run your installation probably did not finish successfully.

> second question, say I fixed some issue on the code, do I have to push another commit? can't I just ammend a patchset to the PR (like Gerrit)?? --- sorry but I am new to github :)

To allow others follow changes it is better to push another commit

jafar-atili commented:
> > Thanks for your reply My development environment is set on a devcontainer using docker and VSC following the instruction you mentioned. However I struggled at the beginning because of the docker came with early version of python 3 and many dependency were not working, I had to manually build python3.9 and install all the deps - hope there's a easy solution for this.
> 
> The dev container comes with it's own python, you don't need to install python on the host system. It looks like something did not work for you on the initial installation.
> 
> Try running `pre-commit`, if it doesn't run your installation probably did not finish successfully.
> 
> > second question, say I fixed some issue on the code, do I have to push another commit? can't I just ammend a patchset to the PR (like Gerrit)?? --- sorry but I am new to github :)
> 
> To allow others follow changes it is better to push another commit

The pre-commit ran and all passed locally, I'll try to understand the error and see what is missing, I'll push another commit

jafar-atili commented:
Hi,

I am trying to perform a new commit, however the pre-commit fails on 2 items that if I run independently it works:

[WARNING] Unstaged files detected.
[INFO] Stashing unstaged files to /home/vscode/.cache/pre-commit/patch1650289626-9311.
pyupgrade................................................................Passed
black....................................................................Passed
codespell................................................................Passed
flake8...................................................................Passed
bandit...................................................................Passed
isort....................................................................Passed
Check JSON...............................................................Passed
Don't commit to branch...................................................Passed
yamllint.............................................(no files to check)Skipped
prettier.................................................................Passed
mypy.....................................................................Passed
pylint...................................................................Passed
**gen_requirements_all.....................................................Failed**
- hook id: gen_requirements_all
- files were modified by this hook
**hassfest.................................................................Failed**
- hook id: hassfest
- files were modified by this hook

Validating codeowners... done in 0.02s
Validating config_flow... done in 0.03s
Validating dependencies... done in 10.08s
Validating dhcp... done in 0.00s
Validating json... done in 0.00s
Validating manifest... done in 0.06s
Validating mqtt... done in 0.00s
Validating requirements... done in 0.01s
Validating services... done in 3.35s
Validating ssdp... done in 0.00s
Validating translations... done in 1.77s
Validating usb... done in 0.00s
Validating zeroconf... done in 0.00s
Validating coverage... done in 0.04s
Validating mypy_config... done in 0.03s
Validating metadata... done in 0.00s

Integrations: 1016
Invalid integrations: 0

hassfest-metadata....................................(no files to check)Skipped
hassfest-mypy-config.................................(no files to check)Skipped
[WARNING] Stashed changes conflicted with hook auto-fixes... Rolling back fixes...
[INFO] Restored changes from /home/vscode/.cache/pre-commit/patch1650289626-9311


If I run 

python3 -m script.hassfest
Validating codeowners... done in 0.03s
Validating config_flow... done in 0.04s
Validating dependencies... done in 10.47s
Validating dhcp... done in 0.01s
Validating json... done in 0.00s
Validating manifest... done in 0.08s
Validating mqtt... done in 0.00s
Validating requirements... done in 0.02s
Validating services... done in 3.22s
Validating ssdp... done in 0.00s
Validating translations... done in 4.35s
Validating usb... done in 0.00s
Validating zeroconf... done in 0.00s
Validating coverage... done in 0.16s
Validating mypy_config... done in 0.04s
Validating metadata... done in 0.01s

Integrations: 1016
Invalid integrations: 0

$ python3 -m script.gen_requirements_all  
$ echo $?
0

can you assist?

jafar-atili committed:
improved API and more logs

jafar-atili committed:
improved API and more logs

jafar-atili committed:
fixed test_config_flow code

jafar-atili committed:
fixed test_config_flow code

thecode commented:
Since this is a new integration, please limit your PR to one platform, I suggest to start with the `Switch` platform.

jafar-atili committed:
removed light and cover

jafar-atili committed:
removed light and cover

jafar-atili commented:
I still getting the test fails in the CI, I am unable to run tox it takes time and it fails - not sure how to test locally

thecode commented on .coveragerc:
```diff
@@ -1495,13 +1495,16 @@ omit =
 exclude_lines =
     # Have to re-enable the standard pragma
     pragma: no cover
-
```
Please add this link back

thecode commented on homeassistant/components/switchbee/const.py:
```diff
@@ -0,0 +1,7 @@
+"""Constants for the SwitchBee Smart Home integration."""
+
+from datetime import timedelta
+
```
It would be better to keep here only the `SCAN_INTERVAL_SEC` do the calculation to `timedelta` where you use it

thecode replied:
```suggestion
```

thecode commented on homeassistant/components/switchbee/const.py:
```diff
@@ -0,0 +1,7 @@
+
+DOMAIN = "switchbee"
+SCAN_INTERVAL_SEC = 5
+SCAN_INTERVAL = timedelta(seconds=SCAN_INTERVAL_SEC)
```
```suggestion
```

thecode commented on homeassistant/components/switchbee/__init__.py:
```diff
@@ -0,0 +1,137 @@
+        self._devices = None
+        self._mac = ""
+        self._reconnect_counts = 0
+        super().__init__(hass, _LOGGER, name=DOMAIN, update_interval=SCAN_INTERVAL)
```
```suggestion
        super().__init__(hass, _LOGGER, name=DOMAIN, update_interval=timedelta(SCAN_INTERVAL_SEC))
```

thecode commented on homeassistant/components/switchbee/__init__.py:
```diff
@@ -0,0 +1,137 @@
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+from .const import DOMAIN, SCAN_INTERVAL
```
```suggestion
from .const import DOMAIN, SCAN_INTERVAL_SEC
```
also make sure to import `timedelta`

thecode commented on homeassistant/components/switchbee/__init__.py:
```diff
@@ -0,0 +1,137 @@
+    api = switchbee.SwitchBee(central_unit, user, password, websession)
+    resp = await api.login()
+    if resp[switchbee.ATTR_STATUS] != switchbee.STATUS_OK:
+        _LOGGER.error(
+            "Failed to login to the central unit (%s) with the user %s %s",
+            central_unit,
+            central_unit,
+            resp,
+        )
```
```suggestion
        _LOGGER.error(
            "Failed to login to the central unit (%s) with the user %s %s",
            central_unit,
            central_unit,
            resp,
        )
```
Please include this message in the `PlatformNotReady` instead of logging error. This will make sure it is only logged once.
e.g. `raise PlatformNotReady("some error")`

thecode commented on homeassistant/components/switchbee/__init__.py:
```diff
@@ -0,0 +1,137 @@
+
+    def __init__(self, hass, swb_api):
+        """Initialize."""
+        self._hass = hass
```
```suggestion
```
`self.hass` is already assigned by `DataUpdateCoordinator`  class

thecode commented on homeassistant/components/switchbee/__init__.py:
```diff
@@ -0,0 +1,137 @@
+    async def _async_update_data(self):
+        if self._reconnect_counts != self._api.reconnect_count:
+            self._reconnect_counts = self._api.reconnect_count
+            _LOGGER.info(
```
```suggestion
            _LOGGER.debug(
```
Try to minimize using `info` log level

thecode commented on homeassistant/components/switchbee/__init__.py:
```diff
@@ -0,0 +1,137 @@
+
+        if self._devices is None:
+            result = await self._api.get_configuration()
+            _LOGGER.info("Loaded devices")
```
debug level

thecode commented on homeassistant/components/switchbee/config_flow.py:
```diff
@@ -0,0 +1,98 @@
+class Hub:
+    """Placeholder class to make tests pass.
+
+    TODO Remove this placeholder class and replace with things from your PyPI package.
```
`TODO`

jafar-atili committed:
Fixed CR comments, updated pylib, improved response time and lowered the scan interval for lower latency

jafar-atili committed:
Fixed CR comments, updated pylib, improved response time and lowered the scan interval for lower latency

thecode commented on homeassistant/components/switchbee/__init__.py:
```diff
@@ -0,0 +1,138 @@
+                    ]:
+                        self._devices[item[switchbee.ATTR_ID]] = item
+
+        if self._devices is None:
+            raise UpdateFailed("Failed to fetch devices")
```
```suggestion
            if self._devices is None:
                raise UpdateFailed("Failed to fetch devices")
```
Move inside the first `if` to avoid running this check on every update

thecode commented on homeassistant/components/switchbee/switch.py:
```diff
@@ -0,0 +1,100 @@
+        self._attr_unique_id = device["uid"]
+        self._attr_device_info = DeviceInfo(
+            identifiers={
+                (DOMAIN, device["uid"]),
```
If you use the `uid` as device identifier also set it in config flow and check if it is configured already.

jafar-atili replied:
Is this identifier used for the entity? or the integration (hub/device)?

I combined this uid from the CentralUnitMAC-UniqueSwitchID to identify this entity (switch in this case)
Its enough to use the MAC address of the c unit in order to detect conflict

you can't have the same switch/cover/light under the 2 different central units.

jafar-atili commented on homeassistant/components/switchbee/config_flow.py:
```diff
@@ -0,0 +1,95 @@
+)
+
+
+class Hub:
+    """Placeholder class to make tests pass."""
+
+    def __init__(self, central_unit_ip: str) -> None:
+        """Initialize."""
+        self.central_unit_ip = central_unit_ip
+
+    async def authenticate(
+        self, hass: HomeAssistant, username: str, password: str
+    ) -> dict:
+        """Test if we can authenticate with the host."""
+        websession = async_get_clientsession(hass, verify_ssl=False)
+        return await switchbee.SwitchBee(
+            self.central_unit_ip, username, password, websession
+        ).login()
```
are there any clear documentation for writing test? it's not clear to me how it works.

jafar-atili commented:
Not sure if need to click something to publish my review comments... just in case

jafar-atili committed:
CR fixes, added advanced setup form to let the
user choose the following:
    - scan interval in seconds: default 5
    - whether to expose scenarios and group switches from the CU or not

jafar-atili committed:
CR fixes, added advanced setup form to let the
user choose the following:
    - scan interval in seconds: default 5
    - whether to expose scenarios and group switches from the CU or not

jafar-atili committed:
used SCAN_INTERVAL_SEC instead of typing just the number

jafar-atili committed:
used SCAN_INTERVAL_SEC instead of typing just the number

thecode commented on homeassistant/components/switchbee/config_flow.py:
```diff
@@ -33,7 +38,7 @@
+        vol.Required(CONF_SCAN_INTERVAL, default=SCAN_INTERVAL_SEC): cv.positive_int,
         vol.Required(CONF_EXPOSE_SCENARIOS, default=False): cv.boolean,
         vol.Required(CONF_EXPOSE_GROUP_SWITCHES, default=False): cv.boolean,
     }
```
These should not be set via config flow, but via config options https://developers.home-assistant.io/docs/config_entries_options_flow_handler

I highly suggest to focus on making the PR small as possible and try not to add additional functionality during review (unless this is needed to fix something raised in the review). 

Thanks 👍

jafar-atili commented on homeassistant/components/switchbee/__init__.py:
```diff
@@ -0,0 +1,138 @@
+            update_interval=timedelta(seconds=SCAN_INTERVAL_SEC),
+        )
+
+    @property
+    def api(self):
+        """Return SwitchBee API object."""
+        return self._api
```
I understand your statement about the content, this is all of the initial content I want this integration to have, and I do prefer to have it from day one so it get testing coverage as mush as possible (I am already working with this integration)

jafar-atili committed:
Fixed PR comments, added unit tests

jafar-atili committed:
Fixed PR comments, added unit tests

jafar-atili commented:
@thecode, I have completed the unit-tests module, also migrated the advanced config to be optional, all tested together locally and seems to be working fine.

jafar-atili commented:
@thecode, I just wonder if this is pending on me to do changes or just taking time due to the backlog.

thecode commented:
It is pending for someone to have time to review, in the meantime I suggest to fix the failing tests. config flow must have 100% tests coverage and there is still an open issue with adding the same hub again https://github.com/home-assistant/core/pull/70201#discussion_r853151699

jafar-atili commented:
I added a functionality that aborts the config flow if the same hub already configured (based on the mac), and in the last commit I uploaded the config_flow tests.

There is one test that fails in the CI (Python 3.10) but the reason is not clear.

jafar-atili committed:
fixes

jafar-atili committed:
fixes

jafar-atili committed:
Improved the pypi and updated the code accordingly

jafar-atili committed:
Improved the pypi and updated the code accordingly

jafar-atili committed:
fixes

jafar-atili committed:
Merge branch 'switchbee-integration' of https://github.com/jafar-atili/core into switchbee-integration

homeassistant commented:
Hi @GitHub-Action, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#70201).

Once you do that we will be able to review and accept this pull request.

Thanks!

jafar-atili committed:
Merge branch 'dev' into switchbee-integration

jafar-atili committed:
restored new line in .coveragerc

jafar-atili committed:
test: increased config_flow test coverage

jafar-atili commented:
https://github.com/home-assistant/home-assistant.io/pull/24019

jafar-atili committed:
removed two way type

jafar-atili committed:
Updated CODEOWNERS

thecode commented on homeassistant/components/switchbee/config_flow.py:
```diff
@@ -0,0 +1,150 @@
+
+        data_schema = {
+            vol.Required(
+                CONF_SCAN_INTERVAL,
+                default=self.config_entry.options.get(
+                    CONF_SCAN_INTERVAL, SCAN_INTERVAL_SEC
+                ),
+            ): cv.positive_int,
```
It is not allowed to let the user set the scan interval, the intergration should provide a sane default and if the user wants to customize this he can disable polling and make an automation that refresh data.

thecode commented on homeassistant/components/switchbee/__init__.py:
```diff
@@ -0,0 +1,160 @@
+        return False
+
+    coordinator = SwitchBeeCoordinator(
+        hass, api, scan_interval, allowed_devices, entry.data[CONF_SWITCHES_AS_LIGHTS]
```
```suggestion
        hass, api, SCAN_INTERVAL_SEC, allowed_devices, entry.data[
            CONF_SWITCHES_AS_LIGHTS
        ]
```

thecode commented on homeassistant/components/switchbee/__init__.py:
```diff
@@ -0,0 +1,160 @@
+        devices_map[device]
+        for device in entry.options.get(CONF_DEVICES, CONF_DEFUALT_ALLOWED)
+    ]
+    scan_interval = entry.options.get(CONF_SCAN_INTERVAL, SCAN_INTERVAL_SEC)
```
See comment in config flow, this should be fixed, set it from the const and not from config flow
```suggestion
```

thecode commented on tests/components/switchbee/test_data.py:
```diff
@@ -0,0 +1,141 @@
+"""Test data for the tests of the SwitchBee integration."""
```
A better option would be to move this content to `__init__.py` and remove this file

jafar-atili committed:
fix: code review comments

jafar-atili commented:
Documentation:
https://github.com/home-assistant/home-assistant.io/pull/24019

thecode commented on tests/components/switchbee/test_config_flow.py:
```diff
@@ -0,0 +1,202 @@
+        )
+        await hass.async_block_till_done()
+
+    print(result2)
```
```suggestion
```
Prints are not allowed

thecode commented on tests/components/switchbee/test_config_flow.py:
```diff
@@ -0,0 +1,202 @@
+from homeassistant.data_entry_flow import RESULT_TYPE_FORM, FlowResultType
+
+from tests.common import MockConfigEntry
+from tests.components.switchbee import (
```
```suggestion
from . import (
```
Use relative imports

thecode commented on homeassistant/components/switchbee/const.py:
```diff
@@ -0,0 +1,14 @@
+
+DOMAIN = "switchbee"
+SCAN_INTERVAL_SEC = 5
+CONF_SCAN_INTERVAL = "scan_interval"
```
```suggestion
```
Remove leftover

frenck replied:
This comment was marked as resolved, but actually not resolved.

thecode commented on homeassistant/components/switchbee/switch.py:
```diff
@@ -0,0 +1,127 @@
+        self._attr_available = True
+        self._attr_has_entity_name = True
+
+    @property
+    def available(self) -> bool:
+        """Available."""
+        return self._attr_available
+
```
```suggestion
```
This is not needed since you set `self._attr_available` in `_handle_coordinator_update`

jafar-atili replied:
without it the making the device unavailable didn't work

thecode commented on homeassistant/components/switchbee/translations/en.json:
```diff
@@ -0,0 +1,32 @@
+                  "password": "Password",
+                  "switch_as_light": "Initialize switches as light entities"
+              },
+              "title": "Setup SwitchBee Central Unit"
```
```suggestion
```
Remove

thecode commented on homeassistant/components/switchbee/translations/en.json:
```diff
@@ -0,0 +1,32 @@
+      },
+      "step": {
+          "user": {
+              "data": {
```
```suggestion
              "description": "Setup SwitchBee integration with Home Assistant.",
              "data": {
```
Description is missing:
https://github.com/home-assistant/core/blob/3d5fcbe634902ae6795b57784288d2f26b2b64a6/homeassistant/components/switchbee/strings.json#L5

thecode commented on homeassistant/components/switchbee/strings.json:
```diff
@@ -0,0 +1,31 @@
+        "data": {
+          "host": "[%key:common::config_flow::data::host%]",
+          "username": "[%key:common::config_flow::data::username%]",
+          "password": "[%key:common::config_flow::data::password%]"
```
```suggestion
          "password": "[%key:common::config_flow::data::password%]",
          "switch_as_light": "Initialize switches as light entities"
```
Missing:
https://github.com/home-assistant/core/blob/3d5fcbe634902ae6795b57784288d2f26b2b64a6/homeassistant/components/switchbee/translations/en.json#L17

thecode reviewed (changes requested):
Minor final changes

jafar-atili committed:
fixed review comments

jafar-atili committed:
added device_info

jafar-atili commented:
https://github.com/home-assistant/brands/pull/3651

jafar-atili committed:
moved device info to attribute

thecode reviewed (approved):
Looks good, Thanks @jafar-atili

frenck commented on homeassistant/components/switchbee/config_flow.py:
```diff
@@ -0,0 +1,133 @@
+        vol.Required(CONF_HOST): cv.string,
+        vol.Required(CONF_USERNAME): cv.string,
+        vol.Required(CONF_PASSWORD): cv.string,
+        vol.Required(CONF_SWITCHES_AS_LIGHTS, default=False): cv.boolean,
```
This seems to be weird? Home Assistant already provides this functionality out of the box? Why does this integration implement its own logic?

jafar-atili replied:
CONF_SWITCHES_AS_LIGHTS will be removed

frenck commented on homeassistant/components/switchbee/config_flow.py:
```diff
@@ -0,0 +1,133 @@
+            DeviceType.TimedPowerSwitch,
+        ]
+
+        data_schema = {
+            vol.Required(
+                CONF_DEVICES,
+                default=self.config_entry.options.get(
+                    CONF_DEVICES,
+                    CONF_DEFUALT_ALLOWED,
+                ),
+            ): cv.multi_select([device.display for device in all_devices]),
+        }
```
Device selection / inclusion/exclusion is not a task of an integration. It should provide whatever it can. Users can enable/disable what they like via the Home Assistant interface already. Please remove this.

frenck commented on homeassistant/components/switchbee/__init__.py:
```diff
@@ -0,0 +1,156 @@
+    await hass.config_entries.async_reload(config_entry.entry_id)
+
+
+class SwitchBeeCoordinator(DataUpdateCoordinator):
```
Consider moving the coordinator into a separate module (`coordinator.py`), which is the standard.

frenck commented on homeassistant/components/switchbee/__init__.py:
```diff
@@ -0,0 +1,156 @@
+    ]
+    websession = async_get_clientsession(hass, verify_ssl=False)
+    api = CentralUnitAPI(central_unit, user, password, websession)
+    try:
+        await api.connect()
+    except SwitchBeeError:
+        return False
```
Shouldn't it retry?

frenck commented on homeassistant/components/switchbee/__init__.py:
```diff
@@ -0,0 +1,156 @@
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up SwitchBee Smart Home from a config entry."""
+    hass.data.setdefault(DOMAIN, {})
+    central_unit = entry.data[CONF_HOST]
+    user = entry.data[CONF_USERNAME]
+    password = entry.data[CONF_PASSWORD]
+    devices_map: dict[str, DeviceType] = {s.display: s for s in DeviceType}
+    allowed_devices = [
+        devices_map[device]
+        for device in entry.options.get(CONF_DEVICES, CONF_DEFUALT_ALLOWED)
+    ]
+    websession = async_get_clientsession(hass, verify_ssl=False)
+    api = CentralUnitAPI(central_unit, user, password, websession)
```
All this stuff can be moved into the coordinator, that can set it up on first refresh.

frenck commented on homeassistant/components/switchbee/__init__.py:
```diff
@@ -0,0 +1,156 @@
+    coordinator = SwitchBeeCoordinator(
+        hass,
+        api,
+        SCAN_INTERVAL_SEC,
```
Passing in a constant into the constructor of a local class is kinda weird. Just use the constant in the destination directly.

frenck commented on homeassistant/components/switchbee/__init__.py:
```diff
@@ -0,0 +1,156 @@
+        hass,
+        api,
+        SCAN_INTERVAL_SEC,
+        allowed_devices,
+        entry.data[CONF_SWITCHES_AS_LIGHTS],
```
Instead of passing all these things, you could also pass the entry to the coordinator (so it can figure it out itself).

frenck commented on homeassistant/components/switchbee/__init__.py:
```diff
@@ -0,0 +1,156 @@
+    entry.async_on_unload(entry.add_update_listener(update_listener))
+    hass.data[DOMAIN][entry.entry_id] = coordinator
+
+    hass.config_entries.async_setup_platforms(entry, PLATFORMS)
```
This is a bug/legacy

```py
    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
```

frenck commented on homeassistant/components/switchbee/switch.py:
```diff
@@ -0,0 +1,141 @@
+            _LOGGER.error(
+                "Failed to set %s state %s, error: %s", self._attr_name, state, exp
+            )
+            self._async_write_ha_state()
```
Why write state when an error occurs? What state is there to write?

frenck commented on homeassistant/components/switchbee/switch.py:
```diff
@@ -0,0 +1,141 @@
+    async def _async_set_state(self, state):
+        try:
+            await self.coordinator.api.set_state(self._device_id, state)
+        except (SwitchBeeError, SwitchBeeDeviceOfflineError) as exp:
```
We want to raise if an service call failed. Please raise an `HomeAssistantError`.

The code currently masks issues and doesn't cause automation to fail. If something now doesn't work, the user has no clue why.

frenck commented on homeassistant/components/switchbee/switch.py:
```diff
@@ -0,0 +1,141 @@
+
+        async def async_refresh_state():
+
+            try:
+                await self.coordinator.api.set_state(self._device_id, "dummy")
+            except SwitchBeeDeviceOfflineError:
+                return
+            except SwitchBeeError:
+                return
```
Why is this done?

jafar-atili replied:
This is a hack to refresh a state of device which was lost (-1)
The device will keep reporting -1 state (even if it come back online) until something trigger it.

frenck commented on homeassistant/components/switchbee/switch.py:
```diff
@@ -0,0 +1,141 @@
+            except SwitchBeeError:
+                return
+
+        if self.coordinator.data[self._device_id].state == -1:
+            # This specific call will refresh the state of the device in the CU
+            self.hass.async_create_task(async_refresh_state())
+
+            if self.available:
+                _LOGGER.error(
+                    "%s switch is not responding, check the status in the SwitchBee mobile app",
+                    self.name,
+                )
+            self._attr_available = False
+            self.async_write_ha_state()
+            return None
+
+        if not self.available:
+            _LOGGER.info(
+                "%s switch is now responding",
+                self.name,
+            )
+        self._attr_available = True
+
+        # timed power switch state will represent a number of minutes until it goes off
+        # regulare switches state is ON/OFF
+        self._attr_is_on = (
+            self.coordinator.data[self._device_id].state != ApiStateCommand.OFF
+        )
+
+        super()._handle_coordinator_update()
```
This is all coordinator data already present in memory and can just be implemented in the property methods directly. No need to override `_handle_coordinator_update` here.

jafar-atili replied:
Sorry,  I don't understand what do you mean, can you refer me to some example?

frenck reviewed (commented):
A couple of comments from a quick scan. Please adjusts these in a new PR.

MartinHjelmare commented on homeassistant/components/switchbee/switch.py:
```diff
@@ -0,0 +1,141 @@
+                )
+            self._attr_available = False
+            self.async_write_ha_state()
+            return None
```
Incorrect return value. Just `return`.

MartinHjelmare commented on homeassistant/components/switchbee/switch.py:
```diff
@@ -0,0 +1,141 @@
+    async def async_added_to_hass(self) -> None:
+        """When entity is added to hass."""
+        await super().async_added_to_hass()
+        self._handle_coordinator_update()
```
Please break out the attribute updates from `_handle_coordinator_update` into a separate method. We want to avoid updating state more than once when adding the entity. The entity state will update once after it has been added by default.

jafar-atili replied:
so if I understood you correctly, I should have my own ```update_entity_attributes``` method which basically update the entity attributes from the coordinator which should be called by ```_handle_coordinator_update``` and  ```async_added_to_hass```?

jafar-atili commented on homeassistant/components/switchbee/switch.py:
```diff
@@ -0,0 +1,141 @@
+        async def async_refresh_state():
+
+            try:
+                await self.coordinator.api.set_state(self._device_id, "dummy")
```
The coordinator fetches the data from the Central Unit and throw it as is to the entity, the entity responsible for taking care of the new status.

I don't see how I can migrate this code out of the entity to the coordinator... I'd be happy for example

MartinHjelmare replied:
If the device lost state is not specific to the switch platform, the coordinator should probably check for this case for all devices and try to set the dummy state before returning the device state and notifying the entities. The entities can just check their state and set unavailable if needed.

jafar-atili replied:
Setting the dummy state will be relevant for the next polling.

If the coordinator start refreshing the state of the offline devices, wouldn't this delay the state update for the platforms?

Bottom line I think you are right and this is more Coordinator work than platform and I will address it.

The question if can I address it in a PR post https://github.com/home-assistant/core/pull/78412?

I think I have already overloaded https://github.com/home-assistant/core/pull/78412 with changes

MartinHjelmare replied:
It's ok to do in a separate PR. 👍

Some thoughts: Does the set state api call take a long time to complete? How common is this problem? Can we gather the calls and await them concurrently? Creating a task for the api calls will allow the entities to update state faster but may complicate the coordinator if we need to manage these tasks to not have them lingering before the next task is created. We can discuss more in the PR when it's up.

jafar-atili replied:
Few milliseconds, this happens in case of the Central Unit (Hub) lost rf communication with the wall outlet (unlikely to happen in a steady state) or of the outlet power was lost (also unlikely)

I think it's possible to gather up all the devices that lost communication andcrun task concurrently but its to much complications for a scenario that unlikely to happen.

Sure, I'll make sure to mention you so we come up with efficient solution.

Thanks a lot!

jafar-atili commented:
@MartinHjelmare, your comments were addressed in https://github.com/home-assistant/core/pull/78412
